ldpd/lde.c

```diff
@@ -2271,3 +2275,104 @@ lde_check_filter_af(int af, struct ldpd_af_conf *af_conf,
 	if (strcmp(af_conf->acl_label_expnull_for, filter_name) == 0)
 		lde_change_expnull_for_filter(af);
 }
+
+void lde_route_update(struct iface *iface)
+{
+	struct fec	*f;
+	struct fec_node	*fn;
+	struct fec_nh	*fnh;
+	struct lde_nbr  *ln;
+
+	/* update label of non-connected routes */
+	RB_FOREACH(f, fec_tree, &ft) {
+		fn = (struct fec_node *)f;
+		if (IS_MPLS_UNRESERVED_LABEL(fn->local_label))
+			continue;
+
+		LIST_FOREACH(fnh, &fn->nexthops, entry) {
+			/*
+			 * If connected leave existing label. If LDP
+			 * configured on interface or a static route
+			 * may need new label. If no LDP configured
+			 * treat fec as a connected route
+			 */
+			if (fnh->flags & F_FEC_NH_CONNECTED)
+				break;
+
+			if (fnh->ifindex == iface->ifindex) {
+				fnh->flags &= ~F_FEC_NH_NO_LDP;
+				if (IS_MPLS_RESERVED_LABEL(fn->local_label)) {
+					fn->local_label = NO_LABEL;
+					fn->local_label = lde_update_label(fn);
+					if (fn->local_label != NO_LABEL)
+						RB_FOREACH(ln, nbr_tree,
+						    &lde_nbrs)
+							lde_send_labelmapping(
+							    ln, fn, 0);
+				}
+				break;
+			}
+		}
+	}
+	RB_FOREACH(ln, nbr_tree, &lde_nbrs)
+		lde_imsg_compose_ldpe(IMSG_MAPPING_ADD_END, ln->peerid,                            0, NULL, 0);
+}
+
+void lde_route_update_release(struct iface *iface, int af)
+{
+	struct lde_nbr	*ln;
+	struct fec	*f;
+	struct fec_node	*fn;
+	struct fec_nh	*fnh;
+
+	/* update label of interfaces no longer running LDP */
+	RB_FOREACH(f, fec_tree, &ft) {
+		fn = (struct fec_node *)f;
+
+		switch (af) {
+		case AF_INET:
+			if (fn->fec.type != FEC_TYPE_IPV4)
+				continue;
+			break;
+		case AF_INET6:
+			if (fn->fec.type != FEC_TYPE_IPV6)
+				continue;
+			break;
+		default:
+			fatalx("lde_route_update_release: unknown af");
+		}
+
+		LIST_FOREACH(fnh, &fn->nexthops, entry) {
+			/*
+			 * If connected leave existing label. If LDP
+			 * removed from interface may need new label
+			 * and would be treated as a connected route
+			 */
+			if (fnh->flags & F_FEC_NH_CONNECTED)
+				break;
+
+			if (fnh->ifindex == iface->ifindex) {
+				fnh->flags |= F_FEC_NH_NO_LDP;
+				if (IS_MPLS_UNRESERVED_LABEL(fn->local_label))
+				{
```

Comment:
nit: opening curly braces should never go on their own line

Reply:
line went past 80 characters with the { at the end of the line so I moved it to the next line

Reply:
I'll add a continue to reduce the indentations.